plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/Prism/StateMachine.hs

```diff
@@ -0,0 +1,120 @@
+{-# LANGUAGE DataKinds          #-}
+{-# LANGUAGE DeriveAnyClass     #-}
+{-# LANGUAGE DeriveGeneric      #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE NamedFieldPuns     #-}
+{-# LANGUAGE NoImplicitPrelude  #-}
+{-# LANGUAGE TemplateHaskell    #-}
+{-# LANGUAGE TypeApplications   #-}
+-- | State machine that manages credential tokens
+module Language.PlutusTx.Coordination.Contracts.Prism.StateMachine(
+    IDState(..)
+    , IDAction(..)
+    , UserCredential(..)
+    , scriptInstance
+    , machineClient
+    , mkMachineClient
+    ) where
+
+import           Data.Aeson                                                (FromJSON, ToJSON)
+import           Data.Hashable                                             (Hashable)
+import           GHC.Generics                                              (Generic)
+import           Language.Plutus.Contract.StateMachine                     (State (..), StateMachine (..),
+                                                                            StateMachineClient (..),
+                                                                            StateMachineInstance (..), Void)
+import qualified Language.Plutus.Contract.StateMachine                     as StateMachine
+import qualified Language.PlutusTx                                         as PlutusTx
+import           Language.PlutusTx.Coordination.Contracts.Prism.Credential (Credential (..), CredentialAuthority (..))
+import qualified Language.PlutusTx.Coordination.Contracts.Prism.Credential as Credential
+import           Language.PlutusTx.Prelude
+import qualified Ledger.Constraints                                        as Constraints
+import           Ledger.Constraints.TxConstraints                          (TxConstraints)
+import           Ledger.Crypto                                             (PubKeyHash)
+import qualified Ledger.Typed.Scripts                                      as Scripts
+import           Ledger.Value                                              (TokenName, Value)
+import qualified Prelude                                                   as Haskell
+
+data IDState =
+    Active -- ^ The credential is active and can be used in transactions
+    | Revoked -- ^ The credential has been revoked and can't be used anymore.
```

Comment:
If it's a state machine per token (which it seems to be), why not just terminate the machine when it's revoked?

Reply:
Ah, that is what happens.